File: tools/cafi/indexer.py
"""Index manager — ties hasher and describer together."""

from __future__ import annotations

import json
import logging
import os
import sys

from cafi.describer import describe_files, triage_files
from cafi.hasher import (
    compute_hash,
    diff_cache,
    load_cache,
    save_cache,
    walk_project,
)

logger = logging.getLogger(__name__)

CACHE_FILENAME = "file-index.json"

DEFAULT_CONFIG = {
    "excludes": [],
    "max_file_size": 102400,
    "concurrency": 3,
    "batch_size": 25,
    "triage": True,
}


class MissingConfigError(Exception):
    """Raised when .claude/.prove.json is not found and CAFI cannot proceed."""


def load_config(project_root: str, require: bool = True) -> dict:
    """Read index config from ``.claude/.prove.json`` under the ``"index"`` key.

    Args:
        project_root: The project root directory.
        require: If True (default), raise ``MissingConfigError`` when
            ``.claude/.prove.json`` is absent. If False, fall back to defaults.

    Returns:
        Config dict with keys ``excludes``, ``max_file_size``, ``concurrency``.
        Falls back to defaults for any missing keys.

    Raises:
        MissingConfigError: If ``require`` is True and ``.claude/.prove.json`` is absent.
    """
    config_path = os.path.join(project_root, ".claude", ".prove.json")
    result = dict(DEFAULT_CONFIG)

    if not os.path.isfile(config_path):
        if require:
            raise MissingConfigError(
                f"No .claude/.prove.json found.\n"
                f"  Project root: {os.path.abspath(project_root)}\n"
                f"  Expected config: {os.path.abspath(config_path)}\n"
                f"CAFI requires a .claude/.prove.json config to run."
            )
        return result

    try:
        with open(config_path, "r") as fh:
            data = json.load(fh)
        index_cfg = data.get("index", {})
        for key in DEFAULT_CONFIG:
            if key in index_cfg:
                result[key] = index_cfg[key]
    except (json.JSONDecodeError, OSError) as exc:
        logger.warning("Could not read config from %s: %s", config_path, exc)

    return result


def _cache_path(project_root: str) -> str:
    """Return the absolute path to the cache file."""
    return os.path.join(project_root, ".prove", CACHE_FILENAME)


def build_index(project_root: str, force: bool = False) -> dict:
    """Run a full or incremental index build.

    1. Load config from ``.claude/.prove.json``
    2. Walk the project to discover files
    3. Compute SHA-256 hashes for each file
    4. Diff against the existing cache
    5. Describe new/stale files via the Claude CLI
    6. Merge results into the cache and persist

    Args:
        project_root: The project root directory.
        force: If True, re-describe every file regardless of cache state.

    Returns:
        Summary dict with counts: ``new``, ``stale``, ``deleted``,
        ``unchanged``, ``total``.
    """
    config = load_config(project_root)
    files = walk_project(
        project_root,
        excludes=config["excludes"],
        max_file_size=config["max_file_size"],
    )

    # Triage: let Claude filter the file list to only index-worthy files
    if config.get("triage", True):
        files = triage_files(files)

    current_hashes: dict[str, str] = {}
    for fp in files:
        full_path = os.path.join(project_root, fp)
        current_hashes[fp] = compute_hash(full_path)

    cache = load_cache(_cache_path(project_root))
    new, stale, deleted, unchanged = diff_cache(current_hashes, cache)

    if force:
        # Treat every current file as needing re-description
        to_describe = list(current_hashes.keys())
        stale = [f for f in current_hashes if f not in new]
        unchanged = []
    else:
        to_describe = new + stale

    # Generate descriptions for new/stale files
    def _progress(done: int, total: int, path: str) -> None:
        print(f"\r  [{done}/{total}] {path}", end="", file=sys.stderr, flush=True)
        if done == total:
            print(file=sys.stderr)

    if to_describe:
        descriptions = describe_files(
            to_describe,
            project_root,
            concurrency=config["concurrency"],
            batch_size=config["batch_size"],
            on_progress=_progress,
        )
    else:
        descriptions = {}

    # Count files that received empty descriptions (CLI failures)
    error_count = sum(1 for fp in to_describe if not descriptions.get(fp))

    # Merge into cache
    cached_files = cache.get("files", {})

    # Remove deleted entries
    for fp in deleted:
        cached_files.pop(fp, None)

    # Update new and stale entries
    for fp in to_describe:
        cached_files[fp] = {
            "hash": current_hashes[fp],
            "description": descriptions.get(fp, ""),
        }

    # Ensure unchanged entries still have their hash up-to-date
    for fp in unchanged:
        if fp in cached_files:
            cached_files[fp]["hash"] = current_hashes[fp]

    cache["files"] = cached_files
    save_cache(_cache_path(project_root), cache)

    return {
        "new": len(new),
        "stale": len(stale),
        "deleted": len(deleted),
        "unchanged": len(unchanged),
        "total": len(current_hashes),
        "errors": error_count,
    }


def get_status(project_root: str) -> dict:
    """Quick status check without running descriptions.

    Args:
        project_root: The project root directory.

    Returns:
        Dict with counts ``new``, ``stale``, ``deleted``, ``unchanged``,
        and ``cache_exists`` boolean.
    """
    config = load_config(project_root)
    files = walk_project(
        project_root,
        excludes=config["excludes"],
        max_file_size=config["max_file_size"],
    )

    current_hashes: dict[str, str] = {}
    for fp in files:
        full_path = os.path.join(project_root, fp)
        current_hashes[fp] = compute_hash(full_path)

    cp = _cache_path(project_root)
    cache = load_cache(cp)
    new, stale, deleted, unchanged = diff_cache(current_hashes, cache)

    return {
        "new": len(new),
        "stale": len(stale),
        "deleted": len(deleted),
        "unchanged": len(unchanged),
        "cache_exists": os.path.isfile(cp),
    }


def get_description(project_root: str, file_path: str) -> str | None:
    """Look up the cached description for a single file.

    Args:
        project_root: The project root directory.
        file_path: Relative path to the file.

    Returns:
        The description string, or None if not in cache.
    """
    cache = load_cache(_cache_path(project_root))
    entry = cache.get("files", {}).get(file_path)
    if entry is None:
        return None
    return entry.get("description")


def clear_cache(project_root: str) -> bool:
    """Delete the cache file.

    Args:
        project_root: The project root directory.

    Returns:
        True if the file existed and was deleted, False otherwise.
    """
    cp = _cache_path(project_root)
    if os.path.isfile(cp):
        os.remove(cp)
        return True
    return False


def lookup(project_root: str, keyword: str) -> list[dict]:
    """Search the file index by keyword, matching against paths and descriptions.

    Args:
        project_root: The project root directory.
        keyword: Search term (case-insensitive). Matches against file paths
                 and routing-hint descriptions.

    Returns:
        List of dicts with ``path`` and ``description`` for matching files,
        sorted by path.
    """
    cache = load_cache(_cache_path(project_root))
    files = cache.get("files", {})
    if not files:
        return []

    keyword_lower = keyword.lower()
    results: list[dict] = []

    for path in sorted(files.keys()):
        entry = files[path]
        desc = entry.get("description", "")
        if keyword_lower in path.lower() or keyword_lower in desc.lower():
            results.append({"path": path, "description": desc})

    return results


def format_index_for_context(project_root: str) -> str:
    """Format all cached descriptions as a compact Markdown block.

    This output is intended for injection into the session context so
    that the LLM agent knows which files exist and when to read them.

    Args:
        project_root: The project root directory.

    Returns:
        Markdown-formatted string, or an empty string if no cache exists.
    """
    cache = load_cache(_cache_path(project_root))
    files = cache.get("files", {})
    if not files:
        return ""

    lines = ["# Project File Index", ""]
    for path in sorted(files.keys()):
        desc = files[path].get("description", "")
        if desc:
            lines.append(f"- `{path}`: {desc}")
        else:
            lines.append(f"- `{path}`: (no description)")

    return "\n".join(lines) + "\n"

File: packages/cli/src/topics/cafi/indexer.ts
/**
 * Index read/query module — config, hashing fast path, and every read-side
 * accessor over `.prove/file-index.json` (status, get, lookup, context,
 * clear). Cache mutation lives in `plan.ts` (additive stat backfill) and
 * `save.ts` (description merge + deletion prune); description generation
 * lives in the driver Claude session, never in this CLI.
 */

import { existsSync, rmSync, statSync } from 'node:fs';
import { join } from 'node:path';
import type { FileCacheEntry } from '@claude-prove/shared';
import { DEFAULT_CONFIG, loadCache, loadToolConfig, walkProject } from '@claude-prove/shared';
import { computeHash, diffCache } from './hasher';

export const CACHE_FILENAME = 'file-index.json';

export function cachePath(projectRoot: string): string {
  return join(projectRoot, '.prove', CACHE_FILENAME);
}

export interface IndexStatus {
  new: number;
  stale: number;
  deleted: number;
  unchanged: number;
  cache_exists: boolean;
}

export interface CafiConfig {
  excludes: string[];
  max_file_size: number;
  batch_size: number;
  triage: boolean;
  [key: string]: unknown;
}

export function loadCafiConfig(projectRoot: string): CafiConfig {
  const raw = loadToolConfig(projectRoot, 'cafi', DEFAULT_CONFIG);
  return {
    excludes: Array.isArray(raw.excludes) ? (raw.excludes as string[]) : [],
    max_file_size: typeof raw.max_file_size === 'number' ? raw.max_file_size : 102400,
    batch_size: typeof raw.batch_size === 'number' ? raw.batch_size : 25,
    triage: raw.triage !== false,
  };
}

/**
 * Hash every file, but skip rehashing when the cached mtime+size match the
 * file on disk. When the fast path fires, the cached hash is reused — this is
 * safe because a matching mtime+size means the inode content has not changed
 * since the last index build. On a miss (new file, or mtime/size changed, or
 * no cached stat), the file is hashed and the new stat is recorded.
 *
 * Residual risk: on filesystems with coarse mtime granularity (HFS+ ticks in
 * whole seconds), a same-size edit within one tick can falsely hit the fast
 * path and reuse a stale hash. Accepted: the cache only feeds description
 * freshness (routing hints), so the worst case is one delayed re-description,
 * and APFS — the common case on darwin — has nanosecond timestamps.
 *
 * Returns both the hash map and the per-file stat used so callers can
 * backfill the stat fields into the cache entry without a second syscall.
 */
export function hashAllFilesWithFastPath(
  projectRoot: string,
  files: string[],
  cachedFiles: Record<string, FileCacheEntry>,
): { hashes: Record<string, string>; stats: Record<string, { mtime_ms: number; size: number }> } {
  const hashes: Record<string, string> = {};
  const stats: Record<string, { mtime_ms: number; size: number }> = {};
  for (const fp of files) {
    const absPath = join(projectRoot, fp);
    const st = statSync(absPath);
    const mtime_ms = st.mtimeMs;
    const size = st.size;
    stats[fp] = { mtime_ms, size };
    const cached = cachedFiles[fp];
    if (
      cached?.mtime_ms !== undefined &&
      cached.size !== undefined &&
      cached.mtime_ms === mtime_ms &&
      cached.size === size
    ) {
      // Fast path: stat fields match — reuse the stored hash without reading file bytes.
      hashes[fp] = cached.hash;
    } else {
      hashes[fp] = computeHash(absPath);
    }
  }
  return { hashes, stats };
}

/**
 * Quick status check without describing files. Mirrors `buildPlan`'s
 * walk + hash + diff phases so the caller sees exactly what a run would do.
 *
 * Cost: "quick" means it skips the (expensive) describe pass, NOT that it is
 * free. Files whose cached mtime+size match on disk reuse their stored hash
 * (stat-only); every other file is synchronously read + SHA-256-hashed on the
 * main thread. Worst case (cold or stale cache) is O(total repo bytes)
 * blocking I/O — effectively a full index walk minus describe. The only bound
 * is the per-file `max_file_size` cap; there is no file-count cap.
 */
export function getStatus(projectRoot: string): IndexStatus {
  const config = loadCafiConfig(projectRoot);
  const files = walkProject(projectRoot, {
    excludes: config.excludes,
    maxFileSize: config.max_file_size,
  });

  const cp = cachePath(projectRoot);
  const cache = loadCache(cp);
  // Use the mtime+size fast path so unchanged files are not re-hashed.
  const { hashes: currentHashes } = hashAllFilesWithFastPath(projectRoot, files, cache.files ?? {});
  const diff = diffCache(currentHashes, cache);

  return {
    new: diff.new.length,
    stale: diff.stale.length,
    deleted: diff.deleted.length,
    unchanged: diff.unchanged.length,
    cache_exists: fileExists(cp),
  };
}

/** Look up the cached description for a single file, or null if absent. */
export function getDescription(projectRoot: string, filePath: string): string | null {
  const cache = loadCache(cachePath(projectRoot));
  const entry = cache.files?.[filePath];
  if (!entry) return null;
  return entry.description;
}

/** Delete the cache file. Returns true if a file was deleted, false if absent. */
export function clearCache(projectRoot: string): boolean {
  const cp = cachePath(projectRoot);
  if (fileExists(cp)) {
    rmSync(cp);
    return true;
  }
  return false;
}

export interface LookupHit {
  path: string;
  description: string;
}

/**
 * Case-insensitive search across cached paths and descriptions. Returns
 * hits sorted by path.
 */
export function lookup(projectRoot: string, keyword: string): LookupHit[] {
  const cache = loadCache(cachePath(projectRoot));
  const files = cache.files ?? {};
  const paths = Object.keys(files);
  if (paths.length === 0) return [];

  const needle = keyword.toLowerCase();
  const hits: LookupHit[] = [];
  for (const path of paths.sort()) {
    const entry = files[path];
    const description = entry?.description ?? '';
    if (path.toLowerCase().includes(needle) || description.toLowerCase().includes(needle)) {
      hits.push({ path, description });
    }
  }
  return hits;
}

/**
 * Render the cache as a compact Markdown block for session-context
 * injection. Returns the empty string when no cache exists.
 */
export function formatIndexForContext(projectRoot: string): string {
  const cache = loadCache(cachePath(projectRoot));
  const files = cache.files ?? {};
  const paths = Object.keys(files);
  if (paths.length === 0) return '';

  const lines: string[] = ['# Project File Index', ''];
  for (const path of paths.sort()) {
    const description = files[path]?.description ?? '';
    if (description) {
      lines.push(`- \`${path}\`: ${description}`);
    } else {
      lines.push(`- \`${path}\`: (no description)`);
    }
  }
  return `${lines.join('\n')}\n`;
}

function fileExists(path: string): boolean {
  if (!existsSync(path)) return false;
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}
